src/k8s/pkg/k8sd/api/certs_refresh.go

```diff
@@ -0,0 +1,346 @@
+			certificates.KubeletCert = string(csr.Status.Certificate)
+		case fmt.Sprintf("k8sd-%d-worker-kubelet-client", req.Seed):
+			certificates.KubeletClientCert = string(csr.Status.Certificate)
+		case fmt.Sprintf("k8sd-%d-worker-kube-proxy-client", req.Seed):
+			certificates.KubeProxyClientCert = string(csr.Status.Certificate)
+		default:
+			log.Error(fmt.Errorf("unknown CSR %s", csrName), "Unknown CSR")
+			return response.InternalError(fmt.Errorf("unknown CSR %s", csrName))
+		}
+
+	}
+
+	// Read the CA and client CA
+	ca, err := os.ReadFile(filepath.Join(snap.KubernetesPKIDir(), "ca.crt"))
+	if err != nil {
+		return response.InternalError(err)
+	}
+	clientCA, err := os.ReadFile(filepath.Join(snap.KubernetesPKIDir(), "client-ca.crt"))
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	// Read the new private keys
+	bytesKubeletKey, err := os.ReadFile(filepath.Join(snap.KubernetesPKIDir(), "kubelet.key.tmp"))
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	bytesKubeletClientKey, err := os.ReadFile(filepath.Join(snap.KubernetesPKIDir(), "kubelet-client.key.tmp"))
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	bytesKubeProxyKey, err := os.ReadFile(filepath.Join(snap.KubernetesPKIDir(), "kube-proxy.key.tmp"))
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	certificates.CACert = string(ca)
+	certificates.ClientCACert = string(clientCA)
+	certificates.KubeletKey = string(bytesKubeletKey)
+	certificates.KubeletClientKey = string(bytesKubeletClientKey)
+	certificates.KubeProxyClientKey = string(bytesKubeProxyKey)
+
+	log.Info("Ensuring worker PKI")
+	if _, err = setup.EnsureWorkerPKI(snap, &certificates); err != nil {
+		return response.InternalError(err)
+	}
+
+	// Kubeconfigs
+	log.Info("Generating new kubeconfigs")
+	if err := setup.Kubeconfig(filepath.Join(snap.KubernetesConfigDir(), "kubelet.conf"), "127.0.0.1:6443", certificates.CACert, certificates.KubeletClientCert, certificates.KubeletClientKey); err != nil {
+		return response.InternalError(fmt.Errorf("failed to generate kubelet kubeconfig: %w", err))
+	}
+	if err := setup.Kubeconfig(filepath.Join(snap.KubernetesConfigDir(), "proxy.conf"), "127.0.0.1:6443", certificates.CACert, certificates.KubeProxyClientCert, certificates.KubeProxyClientKey); err != nil {
+		return response.InternalError(fmt.Errorf("failed to generate kube-proxy kubeconfig: %w", err))
+	}
+
+	// Restart the services
+	log.Info("Restarting kubelet and kube-proxy")
+	if err := snap.RestartService(r.Context(), "kubelet"); err != nil {
+		return response.InternalError(err)
+	}
+	if err := snap.RestartService(r.Context(), "kube-proxy"); err != nil {
+		return response.InternalError(err)
+	}
+
+	// Remove the new private keys
+	log.Info("Removing temporal private keys")
+	if err := os.Remove(filepath.Join(snap.KubernetesPKIDir(), "kubelet.key.tmp")); err != nil {
+		return response.InternalError(err)
+	}
```

Comment:
think: in this instance, we have already signed the certificates, restarted the services, and we only fail to do internal cleanup. this is not an error for the user. also, this is not required at all, see my other comment. The CSRs are to be created during `/run`